File: arena-jface/src/main/java/org/uqbar/lacar/ui/impl/jface/JFaceButtonBuilder.java
package org.uqbar.lacar.ui.impl.jface;

import org.eclipse.core.databinding.observable.value.ComputedValue;
import org.eclipse.core.runtime.IStatus;
import org.eclipse.jface.databinding.swt.SWTObservables;
import org.eclipse.swt.SWT;
import org.eclipse.swt.widgets.Button;

import org.uqbar.lacar.ui.impl.jface.actions.JFaceActionAdapter;
import org.uqbar.lacar.ui.impl.jface.bindings.JFaceBindingBuilder;
import org.uqbar.lacar.ui.model.Action;
import org.uqbar.lacar.ui.model.BindingBuilder;
import org.uqbar.lacar.ui.model.ButtonBuilder;

/**
 * 
 * @author npasserini
 */
public class JFaceButtonBuilder extends JFaceSkinnableControlBuilder<Button> implements ButtonBuilder {
	
	public JFaceButtonBuilder(JFaceContainer context) {
		super(context, new Button(context.getJFaceComposite(), SWT.PUSH));
	}

	public ButtonBuilder setCaption(String caption) {
		this.getWidget().setText(caption);
		return this;
	}

	@Override
	public ButtonBuilder setAsDefault() {
		this.getWidget().getShell().setDefaultButton(this.getWidget());
		return this;
	}

	@Override
	public void disableOnError() {
		new JFaceBindingBuilder(this, SWTObservables.observeEnabled(this.getWidget()), //
			new ComputedValue() {
				@Override
				protected Object calculate() {
					return ((IStatus) JFaceButtonBuilder.this.getContainer().getStatus().getValue()).isOK();
				}
			}).build();
	}

	@Override
	public ButtonBuilder onClick(Action action) {
		this.getWidget().addSelectionListener(new JFaceActionAdapter(this.getContainer(), action));
		return this;
	}
	

	// ********************************************************
	// ** TODO Yet not implemented.
	// ********************************************************

	@Override
	public BindingBuilder observeValue() {
		throw new UnsupportedOperationException("Se intentó observar la propiedad 'value' de un Button, que no tiene dicha propiedad");
	}
}

File: arena-jface/src/main/scala/org/uqbar/lacar/ui/impl/jface/builder/JFaceButtonBuilder.scala
package org.uqbar.lacar.ui.impl.jface.builder

import org.eclipse.core.runtime.IStatus
import org.eclipse.swt.SWT
import org.eclipse.swt.widgets.Button
import org.uqbar.lacar.ui.impl.jface.builder.traits.JFaceContainer
import org.uqbar.lacar.ui.impl.jface.bindings.JFaceBindingBuilder
import org.uqbar.lacar.ui.impl.jface.builder.traits.JFaceClickable
import org.uqbar.lacar.ui.impl.jface.swt.observables.CaptionObservableValue
import org.uqbar.lacar.ui.model.BindingBuilder
import org.uqbar.lacar.ui.model.ButtonBuilder
import org.uqbar.lacar.ui.impl.jface.builder.traits.JFaceWithCaption
import org.uqbar.lacar.ui.impl.jface.builder.traits.WithImageBuilder
import org.eclipse.jface.databinding.swt.SWTObservables
import org.uqbar.lacar.ui.impl.jface.builder.traits.Aesthetic
import org.uqbar.lacar.ui.impl.jface.builder.traits.JFaceSizeable
import org.uqbar.lacar.ui.impl.jface.builder.traits.JFaceEnabledDisabled
import org.uqbar.arena.widgets.traits.WidgetWithAlignment

/**
 * @author jfernandes
 */
class JFaceButtonBuilder(c: JFaceContainer)
  extends JFaceSkinnableControlBuilder[Button](c, new Button(c getJFaceComposite, SWT PUSH))
  //	extends JFaceWidgetBuilder[Button](c, new Button(c getJFaceComposite, SWT PUSH))
  with ButtonBuilder
  with WithImageBuilder[Button]
  with JFaceClickable
  with JFaceWithCaption
  with Aesthetic
  with JFaceSizeable
  with WidgetWithAlignment
  with JFaceEnabledDisabled {

  override def setAsDefault() = {
    widget.getShell.setDefaultButton(widget)
    this
  }

  override def observeValue(): BindingBuilder = throw new UnsupportedOperationException(
      "Se intentó observar la propiedad 'value' de un Button, que no tiene dicha propiedad")

  override def alignLeft() = widget.setAlignment(SWT.LEFT)
  override def alignRight() = widget.setAlignment(SWT.RIGHT)
  override def alignCenter() = widget.setAlignment(SWT.CENTER)
  
}